lib/Analysis/VarEquiv.cpp

```diff
@@ -0,0 +1,392 @@
+    Sets->add(S);
+    Node = add(S , Member);
+  }
+  add(Node->Set, Elem);
+}
+
+bool Partition::isSingleton(Element Elem) const {
+  return NodeMap->get(Elem) == nullptr;
+}
+
+// Make Elem a singleton equivalance class and remove it
+// from any other equivalence classes that is a member of.
+void Partition::makeSingleton(Element Elem) {
+  ListNode *Node = NodeMap->get(Elem);
+  if (Node != nullptr) {
+    Set *S = Node->Set;
+    assert(S != nullptr);
+    unlinkNode(Node);
+    NodeMap->remove(Elem);
+    delete Node;
+    remove_if_trivial(S); // S may point to freed memory after this.
+  }
+}
+
+  // For each equivalence class C with a member in S,
+  // split C into two sets:
+  // * C intersected with S
+  // * C - S.
+  //
+  // S must be a set from a different PartitionEquivalence.
+  // We create a new set for C intersectd with S and use
+  // the original set for S to hold C - S.
+
+void Partition::refine(const Set *S) {
+  Scratch.clear();
+
+  for (ListNode *Current = S->Head; Current != nullptr;
+       Current = Current->Next) {
+    Element CurrentElem = Current->Elem;
+    ListNode *Target = NodeMap->get(CurrentElem);
+    if (Target != nullptr) {
+      Set *TargetSet = Target->Set;
+      Set *Intersected = TargetSet->Intersected;
+      if (Intersected == nullptr) {
+        Intersected = new Set();
+        Sets->add(Intersected);
+        TargetSet->Intersected = Intersected;
+        Scratch.push_back(TargetSet);
+      }
+      moveNode(Intersected, Target);
+    }
+  }
+
+  unsigned count = Scratch.size();
+  for (unsigned i = 0; i < count; i++) {
+    Set *Split = Scratch[i];
+    Set *Intersected = Split->Intersected;
+    Split->Intersected = nullptr;
+    remove_if_trivial(Split);
+    remove_if_trivial(Intersected);
+  }
+  Scratch.clear();
+}
+
+void Partition::refine(const Partition *R) { // TODO: mark R as const?
+  assert(R != this);
+  // First check that all elements of the current equivalence classes
+  // are members of at least one equivalence class in R.  Elements
+  // not in any equivalence class in R are in singleton equivalence
+  // classes.  These are not represented in sets and need to be deleted.
+
+  Sets->assignTo(Scratch);  // avoid modifying list of sets we are iterating over.
+  unsigned count = Scratch.size();
+  for (unsigned i = 0; i < count; i++) {
+    Set *S = Scratch[i];
+    ListNode *Current = S->Head;
+    while (Current != nullptr) {
+      // Save next pointer now in case we delete Current.
+      ListNode *Next = Current->Next;
+      if (R->isSingleton(Current->Elem)) {
+        unlinkNode(Current);
+        NodeMap->remove(Current->Elem);
+        delete Current;
+      }
+      Current = Next;
+    }
+    remove_if_trivial(S); 
+  }
+
+
+  count = R->Sets->size();
+  for (unsigned i = 0; i < count; i++)
+    refine(R->Sets->get(i));
+}
+
+// Return a representative element from the equivalence set
+// for Elem.
+  Element Partition::getRepresentative(Element Elem) const {
```

Comment:
There's spaces in front of the function declaration. I think you accidentally hit tab when creating the function.